### PR TITLE
Expand plugin writer's guide to include RepositoryVersion

### PR DESCRIPTION
**Pull request by werwty**
closes #3220
https://pulp.plan.io/issues/3220

### CODE REVIEW

File: docs/plugin-writer/tasks/add-remove.rst [hunk 1]
```diff
@@ -1,8 +1,27 @@
...
-https://pulp.plan.io/issues/3220
+Repositories have versions. A new respository version is created when its set of content units
+changes.
+
+To facilitate the creation of repository versions a :class:`~pulpcore.plugin.models.RepositoryVersion`
+context manager is provided. Plugin Writers are strongly encouraged to use RepositoryVersion as
+a context manager to provide transactional safety, working directory set up, and cleaning
+up the database on failures::
+
+     with RepositoryVersion.create(repository) as new_version:
```
Comment (daviddavis) [inline docs/plugin-writer/tasks/add-remove.rst]: Maybe add highlighting (ie `.. code-block:: python`)?

File: docs/plugin-writer/tasks/add-remove.rst [hunk 2]
```diff
@@ -1,8 +1,27 @@
...
 
-TODO: RepositoryVersion
-https://pulp.plan.io/issues/3220
+Repositories have versions. A new respository version is created when its set of content units
+changes.
+
+To facilitate the creation of repository versions a :class:`~pulpcore.plugin.models.RepositoryVersion`
+context manager is provided. Plugin Writers are strongly encouraged to use RepositoryVersion as
+a context manager to provide transactional safety, working directory set up, and cleaning
+up the database on failures::
```
Comment (daviddavis) [inline docs/plugin-writer/tasks/add-remove.rst]: s/set up/setup/

also maybe: s/cleaning up the database.../database cleanup after encountering failures/

**Comment by daviddavis** [comment]
Also maybe mention that RepositoryVersions are immutable?

File: docs/plugin-writer/tasks/add-remove.rst [hunk 3]
```diff
@@ -1,8 +1,27 @@
 Adding and Removing Content
 ===========================
 
-TODO: RepositoryVersion
-https://pulp.plan.io/issues/3220
+Repositories have versions. A new respository version is created when its set of content units
```
Comment (daviddavis) [inline docs/plugin-writer/tasks/add-remove.rst]: maybe: /have versions/have immutable versions/

Not sure if that's too technical?

Comment (asmacdo) [reply docs/plugin-writer/tasks/add-remove.rst]: I think "A new immutable respository version is created when its set of content units changes" is sufficient. At the start of the guide, I will point to the overview/concepts of the code docs as prerequisite knowledge.

File: docs/plugin-writer/tasks/add-remove.rst (hunk 1 above)
Comment (asmacdo) [reply docs/plugin-writer/tasks/add-remove.rst]: I think this is redundant. we should just link to this code block in the template.

File: docs/plugin-writer/tasks/add-remove.rst (hunk 3 above)
Comment (werwty) [reply docs/plugin-writer/tasks/add-remove.rst]: Should I clarify what immutable means in this context (can't be changed, but can be deleted)? And add some details about how deleting versions would squash up the changes into the next version..

File: docs/plugin-writer/tasks/add-remove.rst [hunk 4]
```diff
@@ -1,8 +1,27 @@
...
+
+To facilitate the creation of repository versions a :class:`~pulpcore.plugin.models.RepositoryVersion`
+context manager is provided. Plugin Writers are strongly encouraged to use RepositoryVersion as
+a context manager to provide transactional safety, working directory set up, and cleaning
+up the database on failures::
+
+     with RepositoryVersion.create(repository) as new_version:
+
+        # add content manually
+        new_version.add_content(content)
```
Comment (asmacdo) [inline docs/plugin-writer/tasks/add-remove.rst:17]: could we add these options (commented out, if they won't work) to the template code block?

File: docs/plugin-writer/tasks/add-remove.rst (hunk 3 above)
Comment (werwty) [reply docs/plugin-writer/tasks/add-remove.rst]: @asmacdo cool, didn't see your comment. I will do that instead

**Comment by asmacdo** [comment]
I'll merge if this LGTY @daviddavis?

**Review by daviddavis — APPROVED** [review]

**Pull request merged**